global-hooks/migrate/migrate_d8_cni_secret.go

```diff
@@ -0,0 +1,220 @@
+		return fmt.Errorf("cannot init Kubernetes client: %v", err)
+	}
+
+	// skip migration if d8-cni-configuration secret doesn't exist.
+	d8cniSecret, err := kubeCl.CoreV1().Secrets("kube-system").Get(context.TODO(), "d8-cni-configuration", metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		input.LogEntry.Info("d8-cni-configuration secret does not exist, skipping migration")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	moduleConfigs, err := kubeCl.Dynamic().Resource(config.ModuleConfigGVR).List(context.TODO(), metav1.ListOptions{})
+	if err != nil && !errors.IsNotFound(err) {
+		return err
+	}
+
+	for _, mc := range moduleConfigs.Items {
+		if slices.Contains([]string{"cni-cilium", "cni-flannel", "cni-simple-bridge"}, mc.GetName()) {
+			moduleEnabled, exists, err := unstructured.NestedBool(mc.UnstructuredContent(), "spec", "enabled")
+			if err != nil {
+				return err
+			}
+			if !exists {
+				break
+			}
+			if !moduleEnabled {
+				break
+			}
+			input.LogEntry.Infof("Module config for %s found, skipping migration", mc.GetName())
+			return removeD8CniSecret(input, kubeCl, d8cniSecret)
+		}
+	}
+
+	cniBytes, ok := d8cniSecret.Data["cni"]
+	if !ok {
+		return fmt.Errorf("d8-cni-configuration secret does not contain \"cni\" field, skipping migration")
+	}
+
+	cniName := string(cniBytes)
+	cniModuleConfig := &config.ModuleConfig{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       config.ModuleConfigKind,
+			APIVersion: config.ModuleConfigGroup + "/" + config.ModuleConfigVersion,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "cni-" + cniName,
+		},
+	}
+
+	switch cniName {
+	case "simple-bridge":
+		cniModuleConfig.Spec = config.ModuleConfigSpec{
+			Enabled: pointer.Bool(true),
+		}
+	case "flannel":
+		flannelConfig := FlannelConfig{
+			PodNetworkMode: "HostGW",
+		}
+		flannelJSON, ok := d8cniSecret.Data["flannel"]
+		if ok {
+			err := json.Unmarshal(flannelJSON, &flannelConfig)
+			if err != nil {
+				return fmt.Errorf("cannot unmarshal flannel config: %v", err)
+			}
+		}
+		cniModuleConfig.Spec = config.ModuleConfigSpec{
+			Enabled: pointer.Bool(true),
+			Settings: config.SettingsValues{
+				"podNetworkMode": flannelConfig.PodNetworkMode,
+			},
+		}
+		cniModuleConfig.Spec.Version = 1
+
+	case "cilium":
+		ciliumConfig := CiliumConfigStruct{}
+		ciliumJSON, ok := d8cniSecret.Data["cilium"]
+		if ok {
+			err := json.Unmarshal(ciliumJSON, &ciliumConfig)
+			if err != nil {
+				return fmt.Errorf("cannot unmarshal cilium config: %v", err)
+			}
+		}
+
+		ciliumSettings := config.SettingsValues{}
+		switch ciliumConfig.Mode {
+		case "VXLAN":
+			ciliumSettings["tunnelMode"] = "VXLAN"
+		case "Direct":
+			ciliumSettings["tunnelMode"] = "Disabled"
+		case "DirectWithNodeRoutes":
+			ciliumSettings["tunnelMode"] = "Disabled"
+			ciliumSettings["createNodeRoutes"] = true
+		default:
+			return fmt.Errorf("unknown cilium mode %s", ciliumConfig.Mode)
+		}
+
+		ciliumSettings["masqueradeMode"] = ciliumConfig.MasqueradeMode
```

Comment:
Add a condition if ciliumConfig.MasqueradeMode is "".

Reply:
Or above, check if the secret.cilium struct is empty.

Reply:
done